### PR TITLE
[runtime] Transition the trampoline code to use memory managers for memory allocation.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/mini/mini-amd64.c [hunk 1]
```diff
@@ -8661,7 +8661,7 @@ mono_arch_build_imt_trampoline (MonoVTable *vtable, MonoDomain *domain, MonoIMTC
 		size += item->chunk_size;
 	}
 	if (fail_tramp) {
-		code = (guint8 *)mono_method_alloc_generic_virtual_trampoline (domain, size + MONO_TRAMPOLINE_UNWINDINFO_SIZE(0));
+		code = (guint8 *)mono_method_alloc_generic_virtual_trampoline (mono_domain_ambient_memory_manager (domain), size + MONO_TRAMPOLINE_UNWINDINFO_SIZE(0));
```
Comment (lambdageek) [inline mono/mini/mini-amd64.c:8664]: So just to make sure I understand.  Here what we need is a way to go from a `MonoVTable*` to a memory manager, right?

**Review by lambdageek — APPROVED** [review]
Looks good.

There are a lot of calls to get the ambient memory manager - we need to eventually clean those up by plumbing something that knows the ALC, right?

Happy to see the domain code reserve stuff go away.

File: mono/mini/mini-amd64.c (hunk 1 above)
Comment (vargaz) [reply mono/mini/mini-amd64.c:8664]: Probably yes.

**Comment by vargaz** [comment]
Some places basically allocate global memory which is not related to the domain/alc, so those are allocated from the memory manager of the root domain.

**Pull request merged**